scripts/firmware.py

```diff
@@ -0,0 +1,1407 @@
+
+                # Call the appropriate menu directly from the handlers dictionary
+                handlers[self.flash]()
+        else:
+            self.main_menu()
+
+        # Fall back to the main menu if no valid condition is met
+        self.main_menu()
+
+    def set_device(self, uuid: str) -> None:
+        self.selected_device = uuid
+
+    def get_device(self) -> Optional[str]:
+        return self.selected_device  # None if not set, str if set
+
+    def get_firmware(self) -> Optional[str]:
+        return self.selected_firmware
+
+    # Create main menu
+    def main_menu(self):
+        header()
+        self.selected_device = None
+        self.selected_firmware = None
+        menu_items = {
+            1: Menu.MenuItem("Katapult - CAN", self.can.menu),
+            2: Menu.MenuItem("Katapult - USB", self.usb.menu),
+            3: Menu.MenuItem("DFU (Coming Soon)", self.dfu_menu),
+            0: Menu.MenuItem("Exit", lambda: exit()),  # Add exit option explicitly
+        }
+
+        # Create and display the menu
+        menu = Menu("Main Menu", menu_items)
+        menu.display()
+
+    def dfu_menu(self):
+        header()
+        step_title("Select a DFU device")
+
+    def find_firmware_files(
+        self,
+        base_dir: str,
+        search_pattern: str = "*",
+        exclude_pattern: Optional[Union[str, List[str]]] = None,
+        high_temp: bool = False,
+    ) -> List[Tuple[str, str]]:
+        if not os.path.isdir(base_dir):
+            print(f"Base directory does not exist: {base_dir}")
+            return []
+
+        firmware_files: List[Tuple[str, str]] = []
+
+        # Traverse the directory structure
+        for root, _, files in os.walk(base_dir):
+            subdirectory = os.path.relpath(
+                root, base_dir
+            )  # Relative path of the subdirectory
+
+            if high_temp != ("HT" in subdirectory):
+                continue
+
+            for file in files:
+                if file.endswith(".bin"):  # Only process .bin files
+                    if fnmatch.fnmatch(file, search_pattern):  # Match inclusion pattern
+                        # Handle exclude_pattern as a string or list
+                        if exclude_pattern:
+                            if isinstance(exclude_pattern, list):
+                                # Skip files if they match any pattern in the list
+                                if any(
+                                    fnmatch.fnmatch(file, pattern)
+                                    for pattern in exclude_pattern
+                                ):
+                                    continue
+                            elif fnmatch.fnmatch(file, exclude_pattern):
+                                continue
+                        firmware_files.append((subdirectory, file))
+
+        return sorted(firmware_files)  # Sort the results
+
+    def select_latest(self, firmware_files: List[Tuple[str, str]], type: str):

```

Comment:
```suggestion
    def select_latest(self, firmware_files: List[Thingy?], type: str):

```
Can we make `Thingy?` a thing? You can use a `NamedTuple` or `TypedDict`.
I have no idea of what either str is.

Reply:
its (dir, filename) if you look 1 line above :D

Reply:
https://github.com/Cartographer3D/cartographer-klipper/pull/136/commits/e1e140190919043b97681bd7377db21920c23707